### ISSUE TITLE
gstreamer-send example issue

### ISSUE THREAD

**Issue by realL4mbd4** [pions/webrtc#279]
When running gstreamer-send.exe < BROWSER_SDP.txt

The following error occurs:

```
could not allocate udp6 stun:stun.l.google.com:19302: Failed to create STUN client: dial udp6: address stun.l.google.com: no suitable address found

panic: invalid character 'v' looking for beginning of value

goroutine 1 [running]:
github.com/pions/webrtc/examples/util.Check(...)
        C:/msys64/mingw64/src/github.com/pions/webrtc/examples/util/util.go:27
github.com/pions/webrtc/examples/util.Decode(0xc000127800, 0x16f0, 0xc000006040, 0x0, 0x0, 0x5)
        C:/msys64/mingw64/src/github.com/pions/webrtc/examples/util/util.go:78 +0x139
main.main()
        C:/msys64/mingw64/src/github.com/pions/webrtc/examples/gstreamer-send/main.go:51 +0x2aa
```

This error makes sense because in examples/util/util.go on line 78 we try to "unmarshal" the json but the input is not valid json... The code snippet below is the decoded base64 from the browser.

```
v=0
o=- 6475042678492659319 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio video
a=msid-semantic: WMS
```

```panic: invalid character 'v' looking for beginning of value``` is probably caused by ```v=0```

I'd love to help in anyway I can but my knowledge is limited... I would like to help improve the docs for windows a bit. Being my first time developing in Go and my first time trying to cross compile to Windows from Mac I had an insane amount of issues and ended up giving up and then trying to compile natively in Windows was a headache of it's own having never used msys2.... Let me know what I can do to help... gstreamer-send was working a few days ago, I used go get to grab the source code and compile today, that in mind whatever changes were made in the last few days must've propagated to the go get repository (is that pulling directly from the github master??)

THANK YOU FOR BUILDING THIS AWESOME PROJECT 👍

**Comment by Sean-Der** [comment]
Hey @realL4mbd4 

It sounds like you might be using an old version of our webpage to use the demos (which exchanged raw SDP instead of JSON). The right URL is https://jsfiddle.net/Laf7ujeo/164/ 

Would you mind trying again with that URL? thanks!

I am glad that you like pion-WebRTC thank you so much for using and reporting this issue :) if there is anything you would like to see in the future we would love for you to join us in Slack!

**Comment by realL4mbd4** [comment]
Ah, my bad. That is working now, but I'm getting this error with gstreamer:
```
Connection State has changed Checking

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.249: gst_element_get_bus: assertion 'GST_IS_ELEMENT (element)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.250: gst_bus_add_watch_full: assertion 'GST_IS_BUS (bus)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.250: gst_object_unref: assertion 'object != NULL' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.250: gst_bin_get_by_name: assertion 'GST_IS_BIN (bin)' failed

(gstreamer-send.exe:1072): GLib-GObject-CRITICAL **: 18:27:03.250: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(gstreamer-send.exe:1072): GLib-GObject-WARNING **: 18:27:03.250: invalid (NULL) pointer instance

(gstreamer-send.exe:1072): GLib-GObject-CRITICAL **: 18:27:03.250: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.251: gst_element_set_state: assertion 'GST_IS_ELEMENT (element)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.251: gst_element_get_bus: assertion 'GST_IS_ELEMENT (element)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.251: gst_bus_add_watch_full: assertion 'GST_IS_BUS (bus)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.251: gst_object_unref: assertion 'object != NULL' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.251: gst_bin_get_by_name: assertion 'GST_IS_BIN (bin)' failed

(gstreamer-send.exe:1072): GLib-GObject-CRITICAL **: 18:27:03.252: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(gstreamer-send.exe:1072): GLib-GObject-WARNING **: 18:27:03.252: invalid (NULL) pointer instance

(gstreamer-send.exe:1072): GLib-GObject-CRITICAL **: 18:27:03.252: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(gstreamer-send.exe:1072): GStreamer-CRITICAL **: 18:27:03.252: gst_element_set_state: assertion 'GST_IS_ELEMENT (element)' failed
Connection State has changed Connected
```

It's probably something wrong with my configuration... The feeds show up in the jsfiddle but they just load, never actually display the test video. Any ideas?

Thanks for getting back so quick!

**Comment by Sean-Der** [comment]
That sounds like GStreamer is missing some elements. A good debugging tool is to set the environment variable GST_DEBUG. In Linux I usually just run `export GST_DEBUG=*:3` first.

You need `videotestsrc` and `vp8enc` you can use the command line tool `gst-inspect-1.0` to show what elements you do have installed.

**Comment by realL4mbd4** [comment]
I got this output when running with debug:
```
Connection State has changed Checking
0:00:00.089560901  1260        2bca660 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "audiotestsrc"!
0:00:00.089966235  1260        2bca660 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "audiotestsrc"
0:00:00.090178950  1260        2bca660 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "opusenc"!
0:00:00.090400941  1260        2bca660 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "opusenc"
0:00:00.090630970  1260        2bca660 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@0000000000000000]
0:00:00.090683222  1260        2bca660 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "appsink"!
0:00:00.090919125  1260        2bca660 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "appsink"
0:00:00.091217793  1260        2bca660 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@0000000000000000]

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.392: gst_element_get_bus: assertion 'GST_IS_ELEMENT (element)' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.392: gst_bus_add_watch_full: assertion 'GST_IS_BUS (bus)' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.392: gst_object_unref: assertion 'object != NULL' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.393: gst_bin_get_by_name: assertion 'GST_IS_BIN (bin)' failed

(gstreamer-send.exe:1260): GLib-GObject-CRITICAL **: 18:48:29.393: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(gstreamer-send.exe:1260): GLib-GObject-WARNING **: 18:48:29.393: invalid (NULL) pointer instance

(gstreamer-send.exe:1260): GLib-GObject-CRITICAL **: 18:48:29.393: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.393: gst_element_set_state: assertion 'GST_IS_ELEMENT (element)' failed
0:00:00.092921371  1260        2bca660 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "videotestsrc"!
0:00:00.093248173  1260        2bca660 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "videotestsrc"
0:00:00.093298260  1260        2bca660 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "vp8enc"!
0:00:00.093754609  1260        2bca660 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "vp8enc"
0:00:00.094141701  1260        2bca660 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@0000000000000000]
0:00:00.094679673  1260        2bca660 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "appsink"!
0:00:00.095050379  1260        2bca660 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "appsink"
0:00:00.097241224  1260        2bca660 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@0000000000000000]

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.398: gst_element_get_bus: assertion 'GST_IS_ELEMENT (element)' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.398: gst_bus_add_watch_full: assertion 'GST_IS_BUS (bus)' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.398: gst_object_unref: assertion 'object != NULL' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.398: gst_bin_get_by_name: assertion 'GST_IS_BIN (bin)' failed

(gstreamer-send.exe:1260): GLib-GObject-CRITICAL **: 18:48:29.398: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(gstreamer-send.exe:1260): GLib-GObject-WARNING **: 18:48:29.398: invalid (NULL) pointer instance

(gstreamer-send.exe:1260): GLib-GObject-CRITICAL **: 18:48:29.398: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(gstreamer-send.exe:1260): GStreamer-CRITICAL **: 18:48:29.398: gst_element_set_state: assertion 'GST_IS_ELEMENT (element)' failed
```
and this from gst-inspect-1.0
```
WMD@DESKTOP-3I721OP MINGW64 ~
$ gst-inspect-1.0.exe
staticelements:  bin: Generic bin
staticelements:  pipeline: Pipeline object

Total count: 1 plugin, 2 features
```

It looks sparse, I installed pkg-config and gcc using these commands:

```
pacman -S mingw-w64-x86_64-pkg-config
pacman -S mingw-w64-x86_64-gcc
```

It doesn't look like it's finding any of my elements.

**Comment by realL4mbd4** [comment]
We got it!!!

I had an old hacky GST_PLUGIN_SYSTEM_PATH system environmental variable set.  I just went through and deleted all the env vars and it resolved the issues with my plugins not being found.

If I have some time I will enhance the example docs with some more platform specific docs. I'd really like to see some documentation for cross compilation of webrtc+gstreamer+golang, I have a feeling that'll be a real headache but once we get our stuff into production it'll probably be necessary 👍
